File: solutions/lua/bob/1/bob.lua
return {
  hey = function(say)
    say = say:gsub("%s+$", "") -- trim end
    local is_question = say:match("?$")
    local is_shout = say:match('[A-Z]') and not say:match('[a-z]')
    if #say == 0 then return "Fine. Be that way!" end
    if is_question and is_shout then return "Calm down, I know what I'm doing!" end
    if is_question then return "Sure." end
    if is_shout then return "Whoa, chill out!" end
    return "Whatever."
  end
}

File: solutions/go/bob/1/bob.go
// Package bob contains Bob's behaviors.
package bob

import (
	"strings"
	"unicode"
)

func hasLetters(s string) bool {
	for _, c := range s {
		if unicode.IsLetter(c) {
			return true
		}
	}
	return false
}

func allCaps(s string) bool {
	for _, c := range s {
		if unicode.IsLower(c) {
			return false
		}
	}
	return true
}

// Hey is Bob's response generator.
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)

	if hasLetters(remark) && allCaps(remark) {
		if strings.HasSuffix(remark, "?") {
			return "Calm down, I know what I'm doing!"
		}
		return "Whoa, chill out!"
	}

	if strings.HasSuffix(remark, "?") {
		return "Sure."
	}

	if len(remark) == 0 {
		return "Fine. Be that way!"
	}

	return "Whatever."
}
